src/switchboard_oauth.erl

```diff
@@ -0,0 +1,196 @@
+%% @auth Thomas Moulia <jtmoulia@pocketknife.io>
+%% @doc Switchboard's OAuth functions. Switchboard only handles
+%% exchanging refresh tokens for access tokens.
+
+-module(switchboard_oauth).
+
+-export([refresh_to_access_token/1, refresh_to_access_token/3]).
+
+
+%%==============================================================================
+%% Public Interface
+%%==============================================================================
+
+%% @doc Exchange the refresh token for an access token using the
+%% application's `oauth_providers' configuration.
+-spec refresh_to_access_token(imap:auth_xoauth2()) ->
+    {ok, imap:auth_xoauth2()} | {error, _}.
+refresh_to_access_token({xoauth2, _, {_, _}} = Auth) ->
+    %% @todo wrap up the req
+    case oauth_for(Auth) of
+        {ok, Props} ->
+            case get_values([refresh_url,
+                             client_id,
+                             client_secret], Props) of
+                {ok, [RefreshUrl, ClientID, ClientSecret]} ->
+                    refresh_to_access_token(Auth, RefreshUrl,
+                                            {ClientID, ClientSecret});
+                {error, Reason} ->
+                    {error, Reason}
+            end;
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+
+%% @doc Exchange the refresh token for an access token using the provided
+%% configuration.
+-spec refresh_to_access_token(imap:auth_xoauth2(),
+                              binary(),
+                              {binary(), binary()}) ->
+    {ok, imap:auth_xoauth2()} | {error, _}.
+refresh_to_access_token({xoauth2, Account, {RefreshToken, _Provider}},
+                        RefreshUrl,
+                        {ClientID, ClientSecret})
+  when is_binary(ClientID) andalso is_binary(ClientSecret) ->
+    Body = to_query_string([{client_id, ClientID},
+                       {client_secret, ClientSecret},
+                       {refresh_token, RefreshToken},
+                       {grant_type, <<"refresh_token">>}]),
+    ContentType = "application/x-www-form-urlencoded",
+    ReqParams = {binary_to_list(RefreshUrl), [], ContentType, Body},
+    case httpc:request(post, ReqParams, [], []) of
+        {ok, {{_, 200, _}, _, Response}} ->
+            Props = jsx:decode(list_to_binary(Response)),
+            AccessToken = proplists:get_value(<<"access_token">>, Props),
+            {ok, {xoauth2, Account, AccessToken}}
+    end.
+
+
+%%==============================================================================
+%% Internal Functions
+%%==============================================================================
+
+%% @private
+%% @doc Turn a tuple into a query string
+-spec to_query_string(tuple) -> {ok, binary()} | {error, _}.
+to_query_string(Props) ->
```

Comment:
We should add quick unit test for this -- it's only used from `refresh_to_access_token` which isn't tested because of the external dep.